### DISCUSSION TITLE
#求救 虛擬環境建置問題

### DISCUSSION THREAD

**Discussion by lx3211** [1101-GS4719-C/Python-Community-GS4719-C#414] [Q&A]
我建立了一個叫myVenv的虛擬環境
一開始功能正常
可是我昨天檢查的時候
本機跟虛擬環境用pip下載的套件又混在一起了

事實上這已經是第二次了
第一次出問題的時候
我選擇重新建一個虛擬環境
以為是第一次建虛擬環境時出錯
看來是後天失調

如下圖
![](https://i.imgur.com/BRq9HFc.png)

希望有人可以解惑qq

**Comment by JSHT** [comment]
我是用 windows 不確定其他系統是不是也一樣。
如果是使用 conda 的話
可以使用 pip show numpy 看看 package 位置裝在哪嗎?
有可能是在創建虛擬環境的時候沒有指定特別的 python 版本，它就直接綁在當前電腦上沒有綁定環境的 python 版本上了。
如果位置相同的話，可以再創一個指定不同 python 版本的虛擬環境，用 pip 裝套件，一樣用 pip show [package name] 看兩個位置有沒有不同。

**Reply by lx3211** [reply]
我用show看過虛擬跟實際環境中的同樣套件，他們都是同樣指向實際環境的位置( ･᷄ὢ･᷅ )

**Reply by JSHT** [reply]
哇~ 因為我是使用 conda，如果是直接使用預設開啟虛擬環境的 command，會直接連系統中還沒有在 conda 下控管的版本，會有一樣的問題，不過 conda 是一開始就直接綁在一起了，不會一開始可以，後面不行。

**Comment by chia-kai-chang** [comment]
## 安裝virtualenv
語法：`pip install virtualenv`

## 使用virtualenv建立虛擬環境
```
C:\> pip install virtualenv # 安裝虛擬環境的套件virtualenv
C:\> mkdir flask_learning
C:\> cd flask_learning
C:\flask_learning> virtualenv . # .表示在目前的目錄新建虛擬環境
C:\flask_learning> cd Scripts
C:\flask_learning\Scripts> activate # 進入虛擬環境
(flask_learning) C:\flask_learning\Scripts> deactivate # 離開虛擬環境
C:\flask_learning\Scripts>
```

## Quiz
情境：已在 `C:\flask_learning` 建立名為 `flask_learning` 的虛擬環境。現在欲安裝 `flask` 套件，請問會安裝在 `flask_learning` 虛擬環境? 還是會安裝在 Python 的主環境?

|命令列指令|Flask模組安裝的位置|
|--|:--:|
|`(flask_learning) C:\flask_learning\Scripts> pip install flask`|虛擬環境|
|`(flask_learning) C:\flask_learning> pip install flask`|虛擬環境|
|`C:\flask_learning\Scripts> pip install flask`|虛擬環境|
|`C:\flask_learning> pip install flask`|主環境|

**Reply by lx3211** [reply]
老師，我剛建立的時候，在虛擬環境中安裝套件，套件跟本機是分開的，當下有確認過。
是時隔幾日再在虛擬環境安裝套件檢查的時候，才發現實際環境變得跟虛擬環境一樣了。
然而這是第二次發生這個情況了，第一次發生的時候我的處理方式是重建一個新的虛擬環境。
p.s都是在有activate的前提下安裝的

**Comment by frswim** [comment]
有可能是你在安裝套件時忘記先activate，因為我最近也是在用虛擬環境裝有關區塊鏈的套件，並沒有這個問題

**Reply by lx3211** [reply]
有喔 有先啟動環境才安裝的
